tgui/packages/tgui/interfaces/ApcControl.js

```diff
@@ -0,0 +1,335 @@
+          Area
+        </Table.Cell>
+        <Table.Cell collapsing>
+          Charge
+        </Table.Cell>
+        <Table.Cell collapsing textAlign="right">
+          Draw
+        </Table.Cell>
+        <Table.Cell collapsing title="Equipment">
+          Eqp
+        </Table.Cell>
+        <Table.Cell collapsing title="Lighting">
+          Lgt
+        </Table.Cell>
+        <Table.Cell collapsing title="Environment">
+          Env
+        </Table.Cell>
+      </Table.Row>
+    </Section>
+  );
+};
+
+const ApcControlScene = (props, context) => {
+  const { data, act } = useBackend(context);
+
+  const [
+    sortByField,
+  ] = useLocalState(context, 'sortByField', null);
+
+  const apcs = flow([
+    map((apc, i) => ({
+      ...apc,
+      // Generate a unique id
+      id: apc.name + i,
+    })),
+    sortByField === 'name' && sortBy(apc => apc.name),
+    sortByField === 'charge' && sortBy(apc => -apc.charge),
+    sortByField === 'draw' && sortBy(
+      apc => -powerRank(apc.load),
+      apc => -parseFloat(apc.load)),
+  ])(data.apcs);
+  return (
+    <Table>
+      {apcs.map((apc, i) => (
+        <Table.Row header
+          key={apc.id}
+          className="Table__row candystripe">
+          <Table.Cell>
+            <Button
+              icon={apc.operating ? 'power-off' : 'times'}
+              color={apc.operating ? 'good' : 'bad'}
+              onClick={() => act('breaker',
+                { ref: apc.ref }
+              )}
+            />
+          </Table.Cell>
+          <Table.Cell>
+            <Button
+              onClick={() => act('access-apc',
+                { ref: apc.ref }
+              )}>
+              {apc.name}
+            </Button>
+          </Table.Cell>
+          <Table.Cell collapsing>
+            <AreaCharge
+              charging={apc.charging}
+              charge={apc.charge}
+            />
+          </Table.Cell>
+          <Table.Cell collapsing textAlign="right">
+            {apc.load}
+          </Table.Cell>
+          <Table.Cell collapsing>
+            <AreaStatusColorBox
+              target={"equipment"}
+              status={apc.eqp}
+              apc={apc}
+              act={act}
+            />
+          </Table.Cell>
+          <Table.Cell collapsing>
+            <AreaStatusColorBox
+              target={"lighting"}
+              status={apc.lgt}
+              apc={apc}
+              act={act}
+            />
+          </Table.Cell>
+          <Table.Cell collapsing>
+            <AreaStatusColorBox
+              target={"environ"}
+              status={apc.env}
+              apc={apc}
+              act={act}
+            />
+          </Table.Cell>
+        </Table.Row>
+      ))}
+    </Table>
+  );
+};
+
+const LogPanel = (props, context) => {
+  const { data } = useBackend(context);
+
+  const logs = flow([
+    map((line, i) => ({
+      ...line,
+      // Generate a unique id
+      id: line.entry + i,
+    })),
+  ])(data.logs.reverse());
```

Comment:
Just saying that with `flow`, you have an option to structure operations on data as a series of transforms, which is quite nice and makes it easier to reorder or add new transformations when necessary.

Another thing to watch out, is that `Array.prototype.reverse()` is a mutating method, which means, that `data.logs` will become reversed, and not just the return value. By reversing the new object that we return with `map`, you avoid mutating `data.logs`.

```suggestion
  const logs = flow([
    map((line, i) => ({
      ...line,
      // Generate a unique id
      id: line.entry + i,
    })),
    logs => logs.reverse(),
  ])(data.logs);
```

Reply:
That is helpful, thanks